library/core/src/fmt/num.rs

```diff
@@ -458,6 +474,185 @@ mod imp {
+        } else {
+            let d1 = n << 1;
+            *curr -= 2;
+            ptr::copy_nonoverlapping(lut_ptr.offset(d1 as isize), buf_ptr.offset(*curr), 2);
+        }
+    }
+}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl fmt::Display for u128 {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt_u128(*self, true, f)
+    }
+}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl fmt::Display for i128 {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let is_nonnegative = *self >= 0;
+        let n = if is_nonnegative {
+            self.to_u128()
+        } else {
+            // convert the negative num to positive by summing 1 to it's 2 complement
+            (!self.to_u128()).wrapping_add(1)
+        };
+        fmt_u128(n, is_nonnegative, f)
+    }
+}
+
+/// Specialized optimization for u128. Instead of taking two items at a time, it splits
+/// into at most 2 u64s, and then chunks by 10e16, 10e8, 10e4, 10e2, and then 10e1.
+/// It also has to handle 1 last item, as 10^40 > 2^128 > 10^39, whereas
+/// 10^20 > 2^64 > 10^19.
+fn fmt_u128(n: u128, is_nonnegative: bool, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+    // 2^128 is about 3*10^38, so 39 gives an extra byte of space
+    let mut buf = [MaybeUninit::<u8>::uninit(); 39];
+    let mut curr = buf.len() as isize;
+    let buf_ptr = MaybeUninit::first_ptr_mut(&mut buf);
+
+    let (n, rem) = udiv_1e9(n);
+    parse_u64_into(rem, &mut buf, &mut curr);
+
+    if n != 0 {
+        // 0 pad up to point
+        let target = (buf.len() - 19) as isize;
+        // SAFETY: Guaranteed that we wrote at most 19 bytes, and there must be space
+        // remaining since it has length 39
+        unsafe {
+            ptr::write_bytes(buf_ptr.offset(target), b'0', (curr - target) as usize);
+        }
+        curr = target;
+
+        let (n, rem) = udiv_1e9(n);
+        parse_u64_into(rem, &mut buf, &mut curr);
+        // Should this following branch be annotated with unlikely?
+        if n != 0 {
+            let target = (buf.len() - 38) as isize;
+            // SAFETY: At this point we wrote at most 38 bytes, pad up to that point,
+            // There can only be at most 1 digit remaining.
+            unsafe {
+                ptr::write_bytes(buf_ptr.offset(target), b'0', (curr - target) as usize);
+                curr = target - 1;
+                *buf_ptr.offset(curr) = (n as u8) + b'0';
+            }
+        }
+    }
+
+    // SAFETY: `curr` > 0 (since we made `buf` large enough), and all the chars are valid
+    // UTF-8 since `DEC_DIGITS_LUT` is
+    let buf_slice = unsafe {
+        str::from_utf8_unchecked(slice::from_raw_parts(
+            buf_ptr.offset(curr),
+            buf.len() - curr as usize,
+        ))
+    };
+    f.pad_integral(is_nonnegative, "", buf_slice)
+}
+
+/// Partition of `n` into n > 1e19 and rem <= 1e19
+fn udiv_1e9(n: u128) -> (u128, u64) {
```

Comment:
Nit: the name says `1e9`, the implementation does `1e19`.

I find it surprising that LLVM does not strength-reduce the division by constants for 128-bit integers into `shr (mul $CONST1) $CONST2`, like it does for 64-bit ones. I wonder if its actually something infeasible, or if its just them having a wrong conditional in a wrong place.

(If you wanted to look into this http://gmplib.org/~tege/divcnst-pldi94.pdf describes the mechanism to strength-reduce)

Reply:
So, I think the reason this happens is because this algorithm requires the upper half of the multiplication result, which for division by 128 bits means the upper 128-bits of the result when multiplying two 128-bit integers.

Since we don’t have a 256-bit multiplication algorithm, this is not something that's super feasible to implement here.

Reply:
By this algorithm, you mean the division into u128 and u64?
I've taken that component from `itoa`, so I'm afraid my understanding of it it lacking.
After doing some googling to figure out what strength reduction, I'm not sure if this is the place to do such,  as it would seem pretty one-off as you say, but I do wonder if it could be added somewhere upstream.

Reply:
> I'm not sure if this is the place to do such, as it would seem pretty one-off

My problem with this is that once this algorithm lands, it is exceedingly easy to forget to remove this function once something faster, better is implemented upstream. It would be less of an issue if this was as fast as theoretically possible, as even if the upstream (LLVM) is fixed, this code won’t pessimize "just" because it implements the algorithm manually.

Reply:
Would it be alright to update this, not resolve the issues, and leave a marker to update in the future?